design/external-compaction.md

```diff
@@ -0,0 +1,249 @@
+
+![Image of Coordinator Queue](https://github.com/dlmarion/accumulo-website/tree/external-compaction-design/design/coordinator_queue.jpg)
+
+When the coordinator calls getTabletServerQueueInfo() on a tsever via RPC, it would need to take the information returned by the tserver and reconcile it with its global summary of information adding and removing the tserver from priority levels as needed.
+This information could potentially be kept in memory in a structure of the form Map<QueueName, SortedMap<Priority, LinkedHashSet<Tserver>>> and when the coordinator starts it polls tservers and constructs the global queue summary map.  Periodically the coordinator could ask tservers for updated summaries and/or the reserveJob RPC call to tservers could return updated summary information.
+
+
+## Workflow
+
+When the coordinator polls a tserver to get its latest external queue summaries the following could happen:
+
+  1. The tserver returns the latest summary information for all of its external compaction queues.
+  2. The coordinator will add the tserver to priorities that it is not present in in the global queue summary map.
+  3. The coordinator will remove the tserver from priorities that is in in the global queue summary map, but it is not in the summaries just returned by the tserver
+  4. If needed the next tserver pointer in the global queue summary map will be updated.
+
+On request for work from a compactor
+  1. For the queue requested, the coordinator gets the next tserver from the global queue summary map.
+  2. The Coordinator contacts the tablet server to reserve job
+  3. Return compaction job information to compactor
+  4. This must handle concurrent requests properly while minimizing the amount of locking.
+
+## Metrics
+
+TServers will each maintain their own internal queues for external compaction queues with all tablets.  If each tserver emits metrics about their local queues, then sum of all tserver metrics will give information about the external queues across the cluster.
+
+Compactor should emit metrics about the number of compactions running and which queue they are running on.
+
+## Scenarios
+
+Scenario 1
+
+ 1. Tablet T1 has files F1,F4,F5,F6
+ 1. The compaction of files F4,F5,F6 is queued on external executor EE2 for Tablet T1
+ 1. A new file F7 arrives for T1
+ 1. The compaction of F4,F5,F6 on EE2 is canceled and new a compaction of F4,F5,F6,F7 is queued on EE2 for Tablet T1
+ 1. A new file F8 arrives for T1
+ 1. The compaction of F4,F5,F6,F7 on EE2 is canceled and new a compaction of F4,F5,F6,F7,F8 is queued on EE2 for Tablet T1
```

Comment:
Maybe it would be good to have different configurable options for handling this scenario. Option 1 would be to cancel any compactions on the arrival of new files. Option 2 would be to allow the compaction to finish (not cancel on arrival). This would allow users to tune their cluster for faster scans or cleaning old data. This might also be complicated by whether the compaction is a USER or system type.  For example, we may want to allow user compactions to finish vs cancelling a system one.

Reply:
Also, I am not sure if we want to cancel compactions with the arrival of new data. I don't think we currently do that at all.

Reply:
The wording needs to be updated.  The compaction is not yet running and cancel just means its canceled in the queue.  The current code never cancels a running compaction when the plan changes, only queued compactions.

Reply:
Ah OK. So something more like "EE2 is removed from the queue"

Reply:
"Removed from the queue" would be better working.  In the code it actually just cancels it, possibly leaving it in the queue for efficiency.  Something canceled in the queue would never run.  However those are just implementation details, its doing lazy removal.

Below is where only queued compactions are canceled

https://github.com/apache/accumulo/blob/b12d1103d788473168616f28ae87a65bb76aa880/server/tserver/src/main/java/org/apache/accumulo/tserver/compactions/CompactionService.java#L185

Below is where a task is atomically transitioned to cancel, but only if its currently in the queued state.

https://github.com/apache/accumulo/blob/b12d1103d788473168616f28ae87a65bb76aa880/server/tserver/src/main/java/org/apache/accumulo/tserver/compactions/CompactionExecutor.java#L112

Below is where a task atomically transitions from queued state to running.  The task in the canceled state, then the transition will fail.

https://github.com/apache/accumulo/blob/b12d1103d788473168616f28ae87a65bb76aa880/server/tserver/src/main/java/org/apache/accumulo/tserver/compactions/CompactionExecutor.java#L88